Unit Test atomic can not be compiled with CUDA

The test case is not compiling if CUDA is available

```
$ cmake alpaka/test/unit/atomic/
$ make
[ 20%] Building CXX object CMakeFiles/common.dir/alpaka/test/common/src/Dummy.cpp.o
[ 40%] Linking CXX static library libcommon.a
[ 40%] Built target common
[ 80%] Building NVCC (Device) object CMakeFiles/atomic.dir/src/atomic_generated_AtomicTest.cpp.o
[ 80%] Building NVCC (Device) object CMakeFiles/atomic.dir/src/atomic_generated_main.cpp.o
alpaka/include/alpaka/atomic/AtomicCudaBuiltIn.hpp(108) Error: unsupported operation

CMake Error at atomic_generated_AtomicTest.cpp.o.cmake:279 (message):
  Error generating file
  /bigdata/hplsim/scratch/widera/buildAlpaka/CMakeFiles/atomic.dir/src/./atomic_generated_AtomicTest.cpp.o
```

```
module li
Currently Loaded Modulefiles:
  1) gcc/5.3.0              4) cuda/8.0               7) pngwriter/0.7.0
  2) cmake/3.10.1           5) openmpi/2.1.2.cuda80   8) hdf5-parallel/1.8.20
  3) boost/1.62.0           6) zlib/1.2.8             9) libsplash/1.7.0
```

--- a/include/alpaka/atomic/AtomicCudaBuiltIn.hpp
+++ b/include/alpaka/atomic/AtomicCudaBuiltIn.hpp
@@ -277,9 +277,10 @@ namespace alpaka
                     return atomicMin(addr, value);
                 }
             };
+#if BOOST_ARCH_CUDA_DEVICE >= BOOST_VERSION_NUMBER(3, 5, 0)
             //-----------------------------------------------------------------------------
             //! The GPU CUDA accelerator atomic operation.
-            /*template<
+            template<
                typename THierarchy>
             struct AtomicOp<
                 op::Min,
@@ -296,7 +297,8 @@ namespace alpaka
                 {
                     return atomicMin(addr, value);
                 }
-            };*/
+            };
+#endif
 
             //-----------------------------------------------------------------------------
             // Max.
@@ -340,9 +342,10 @@ namespace alpaka
                     return atomicMax(addr, value);
                 }
             };
+#if BOOST_ARCH_CUDA_DEVICE >= BOOST_VERSION_NUMBER(3, 5, 0)
             //-----------------------------------------------------------------------------
             //! The GPU CUDA accelerator atomic operation.
-            /*template<
+            template<
                typename THierarchy>
             struct AtomicOp<
                 op::Max,
@@ -359,7 +362,8 @@ namespace alpaka
                 {
                     return atomicMax(addr, value);
                 }
-            };*/
+            };
+#endif
 
             //-----------------------------------------------------------------------------
             // Exch.
@@ -536,9 +540,10 @@ namespace alpaka
                     return atomicAnd(addr, value);
                 }
             };
+#if BOOST_ARCH_CUDA_DEVICE >= BOOST_VERSION_NUMBER(3, 5, 0)
             //-----------------------------------------------------------------------------
             //! The GPU CUDA accelerator atomic operation.
-            /*template<
+            template<
                 typename THierarchy>
             struct AtomicOp<
                 op::And,
@@ -555,7 +560,8 @@ namespace alpaka
                 {
                     return atomicAnd(addr, value);
                 }
-            };*/
+            };
+#endif
 
             //-----------------------------------------------------------------------------
             // Or.
@@ -600,9 +606,10 @@ namespace alpaka
                     return atomicOr(addr, value);
                 }
             };
+#if BOOST_ARCH_CUDA_DEVICE >= BOOST_VERSION_NUMBER(3, 5, 0)
             //-----------------------------------------------------------------------------
             //! The GPU CUDA accelerator atomic operation.
-            /*template<
+            template<
                typename THierarchy>
             struct AtomicOp<
                 op::Or,
@@ -619,7 +626,8 @@ namespace alpaka
                 {
                     return atomicOr(addr, value);
                 }
-            };*/
+            };
+#endif
 
             //-----------------------------------------------------------------------------
             // Xor.
@@ -664,9 +672,10 @@ namespace alpaka
                     return atomicXor(addr, value);
                 }
             };
+#if BOOST_ARCH_CUDA_DEVICE >= BOOST_VERSION_NUMBER(3, 5, 0)
             //-----------------------------------------------------------------------------
             //! The GPU CUDA accelerator atomic operation.
-            /*template<
+            template<
                typename THierarchy>
             struct AtomicOp<
                 op::Xor,
@@ -683,7 +692,8 @@ namespace alpaka
                 {
                     return atomicXor(addr, value);
                 }
-            };*/
+            };
+#endif
 
             //-----------------------------------------------------------------------------
             // Cas.

--- a/test/unit/atomic/src/AtomicTest.cpp
+++ b/test/unit/atomic/src/AtomicTest.cpp
@@ -56,202 +56,204 @@ public:
         T operandOrig) const
     -> void
     {
+        auto operand = alpaka::block::shared::dyn::getMem<T>(acc);
+
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Add
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Add>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig + value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Sub
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Sub>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig - value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Min
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Min>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = (operandOrig < value) ? operandOrig : value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Max
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Max>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = (operandOrig > value) ? operandOrig : value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Exch
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Exch>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Inc
         {
             // \TODO: Check reset to 0 at 'value'.
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(42);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Inc>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig + 1;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Dec
         {
             // \TODO: Check reset to 'value' at 0.
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(42);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Dec>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig - 1;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::And
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::And>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig & value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Or
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Or>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig | value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Xor
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const value = operandOrig + static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Xor>(
                         acc,
-                        &operand,
+                        operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig ^ value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Cas with match
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const compare = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Cas>(
                         acc,
-                        &operand,
+                        operand,
                         compare,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = value;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Cas without match
         {
-            T operand = operandOrig;
+            *operand = operandOrig;
             T const compare = operandOrig + static_cast<T>(1);
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Cas>(
                         acc,
-                        &operand,
+                        operand,
                         compare,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig;
-            BOOST_VERIFY(operand == reference);
+            BOOST_VERIFY(*operand == reference);
         }
     }
 };
@@ -321,8 +323,10 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(
     TestAtomicOperations<TAcc, std::uint16_t>::testAtomicOperations();
 #endif
     TestAtomicOperations<TAcc, std::int32_t>::testAtomicOperations();
-#if !defined(ALPAKA_CI)
+    // The std::uint32_t test is the only one that is compiled for CUDA
+    // because this is the only type which is supported by all atomic operations.
     TestAtomicOperations<TAcc, std::uint32_t>::testAtomicOperations();
+#if !defined(ALPAKA_CI)
     TestAtomicOperations<TAcc, std::int64_t>::testAtomicOperations();
     TestAtomicOperations<TAcc, std::uint64_t>::testAtomicOperations();
 #endif

--- a/test/unit/atomic/src/AtomicTest.cpp
+++ b/test/unit/atomic/src/AtomicTest.cpp
@@ -56,7 +56,7 @@ public:
         T operandOrig) const
     -> void
     {
-        auto operand = alpaka::block::shared::dyn::getMem<T>(acc);
+        auto operand = alpaka::block::shared::st::allocVar<T, __COUNTER__>(acc);
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Add

--- a/test/unit/atomic/src/AtomicTest.cpp
+++ b/test/unit/atomic/src/AtomicTest.cpp
@@ -56,204 +56,204 @@ public:
         T operandOrig) const
     -> void
     {
-        auto operand = alpaka::block::shared::st::allocVar<T, __COUNTER__>(acc);
+        auto && operand = alpaka::block::shared::st::allocVar<T, __COUNTER__>(acc);
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Add
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Add>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig + value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Sub
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Sub>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig - value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Min
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Min>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = (operandOrig < value) ? operandOrig : value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Max
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Max>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = (operandOrig > value) ? operandOrig : value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Exch
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Exch>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Inc
         {
             // \TODO: Check reset to 0 at 'value'.
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(42);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Inc>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig + 1;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Dec
         {
             // \TODO: Check reset to 'value' at 0.
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(42);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Dec>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig - 1;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::And
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::And>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig & value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Or
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Or>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig | value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Xor
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const value = operandOrig + static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Xor>(
                         acc,
-                        operand,
+                        &operand,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig ^ value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Cas with match
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const compare = operandOrig;
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Cas>(
                         acc,
-                        operand,
+                        &operand,
                         compare,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = value;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
 
         //-----------------------------------------------------------------------------
         // alpaka::atomic::op::Cas without match
         {
-            *operand = operandOrig;
+            operand = operandOrig;
             T const compare = operandOrig + static_cast<T>(1);
             T const value = static_cast<T>(4);
             T const ret =
                 alpaka::atomic::atomicOp<
                     alpaka::atomic::op::Cas>(
                         acc,
-                        operand,
+                        &operand,
                         compare,
                         value);
             BOOST_VERIFY(operandOrig == ret);
             T const reference = operandOrig;
-            BOOST_VERIFY(*operand == reference);
+            BOOST_VERIFY(operand == reference);
         }
     }
 };

